### PR TITLE
Feature/cov

### LINKED ISSUE #6269: cov

### PR DESCRIPTION
**Pull request by Hokchris**
https://github.com/unifyai/ivy/issues/6269

closes #6269

Implemented cov for linear algebra section.

### CODE REVIEW

**Comment by JamieLine** [comment]
Hi! Thanks for the PR!

Most of this looks really good, especially given that its over a large volume of code. There are a few things that need fixing but they don't look too drastic, even though it looks like there are a few it's not because it is bad it's because there is just so much code to work through.

1. Check the logs for the tests on this page, and fix them where they are caused by your code. Linting errors can be fixed by running `flake8` and `black` on your code and it will fix / diagnose the errors. 
2. There are some "test discovery" issues that I found when trying to run the tests locally, but they are just caused by the `tensorflow` backend referencing `np.ndarray` instead of `tf.Tensor`, and a reference in the test to `helpers.dtype_value1_value2_axis` when it shouldn't have that `helpers` section. But these are easy to fix, I got the tests running locally in a few minutes.
3. The tests fail when run with a `torch` backend. This is awkward to spot because the tests don't run with all backends by default. You can specify a `--backend=all` argument to PyTest (depending on your development environment, the exact way you do this will change but I might be able to help if you have difficulty finding how to add this argument) and it'll run against all backends. The test needs to pass with all backends, but from just having a quick look it shouldn't be hard to fix, it looks like once you fix the parameter lists for each backend it should pass.

Overall though I think it looks good. The docstrings look good, the code looks good, the vast majority of what's here is great, there's just a few things that need to be fixed before it can be merged.

Any questions or comments please let me know!

**Comment by Hokchris** [comment]
Hey! I've made a couple changes and they should have fixed issues 1 and 2. I'm not sure if the new "KeyError: __name__" errors are from the torch backend compatibility or if this is a different error entirely. I don't think I saw any other functions with the __name__ defined outright so I'm guessing it's another issue.

**Comment by JamieLine** [comment]
Hi!

That's some great progress! I've had a look and after a bit of investigating I've worked out why the Torch error occurs. Your `torch` backend implementation has the wrong argument list. Basically your backend implementations must all have the same function signature as each other. Then that backend function should "change" the arguments to fit the real backend function. I.e. `ivy.functional.backends.torch.cov` should have the same arguments as `ivy.functional.backends.tensorflow.cov` and the others, and then `ivy.functional.backends.torch.cov` can do whatever is needed to produce the same result as the other backends. I'm not entirely sure why the other backends are all taking 2 arguments when `torch` only takes 1, but I'd imagine there is a way to package the 2 inputs the Ivy API function provides (`x1` and `x2`) to produce the single input that `torch` expects. I also suspect that having 2 inputs in the Ivy API function is probably correct but I haven't had a proper read through the documentation of each framework to work out what the 2 inputs are doing. (I'm not a statistician at all so a lot of those docs go over my head).

In short, change your Torch backend function to have the same parameters as the other backend functions, then make the tests pass again, and ensure that all of the functionality in each backend is available in the Ivy API somehow. There's a section in the docs called Superset Behaviour which goes into more detail.

Hope this helps! If there are any comments or questions feel free to let me know, I've had to write this quite quickly so some things might not be very clear and I'm happy to clarify if that is the case!

**Comment by Hokchris** [comment]
I made some changes to the backends, looks like that might've fixed it. I'm still working on the final part of the tensorflow backend, it was a bit more difficult than I anticipated. I tested it agaist numpy's backend in a Google colab and it seemed to work as intended, just need to add the aweights and fweights functionality to the backend.

**Comment by Hokchris** [comment]
Hey! I made some changes again to most of the files, and I'm getting a new AssertionError when running locally and with the commits. I did a new merge from the master branch and it seemed to fix the old KeyError: "__name__" but now there's an assertion error instead. I saw some other PR requests have the same issue and the issue persists if I skip my own function tests locally so I'm not sure if there's something I should change or wait for a change on the master branch.

**Comment by JamieLine** [comment]
Hi! Sorry this is so late, I've had a few busy days at Uni, I've had a look and I can tell you what is happening here.

So the AssertionError I got was `Hypothesis has found multiple failures` which I'll be honest is a completely unhelpful error message. However, lower down in the output it should show what the errors are, and what examples were used to cause them.

If the specific errors don't appear, add `-s` as an argument to PyTest. Doing this will make the PyTest output very long but it will ensure that everything is output at some point.

The first error I saw (I haven't looked through all of them) is that the call to `torch.vstack` passes in one too many arguments, it should take a list of `Tensor`s. 

The rest don't look too bad to fix from what I have seen, and hopefully this message should give you some direction.

If you need any help finding the specific errors or anything else please let me know!

**Comment by JamieLine** [comment]
Hi! 

I'm glad to see you're still working on the PR, and it's good that you've updated your local branch to match the current master, just wanted to point out that somehow a lot of files have picked up new changes that aren't showing in the master branch, i.e. the files changed tab on this page has a huge number of changes. If that's just GitHub being weird and you haven't made that many changes from master then there's no issue, but I just want to say that before this PR can be merged it needs to only contain the changes necessary for the PRs functionality.

Thanks!

**Comment by Hokchris** [comment]
Hey! I did a force push on the branch so it should just have my commit showing, I think this should've fixed the tensorflow backend for cov.

**Comment by JamieLine** [comment]
Hi!

The other changes have all cleared out of the diff now which is good, and everything looks good other than 2 things, I just ran the test and got 2 errors, the Tensorflow backend gave me an error `AssertionError: the return with a tensorflow backend produced data type of float64, while the return with a jax backend returned a data type of float32.` and the other backends are complaining that they are getting inputs of more than 2 dimensions, both should be easy enough to fix.

Once they are fixed it should be ready to merge!

Thanks!

**Comment by JamieLine** [comment]
Hi! Just wanted to mention that recently some changes were made to how tests work, please update your fork as there have been some bugs that I believe are now fixed (I know most of them have because I fixed them but I don't know if there are some that I missed)

**Comment by JamieLine** [comment]
Unassigning myself as I no longer work for Ivy, all on good terms I just need to focus on my studies. The bot should soon re-allocate someone.

**Comment by Hokchris** [comment]
Hey! I made a small change to one of the parameters in the test file but I'm currently getting a very long list of errors after running through the tests locally in the tests/functional/core/linalg file. Not sure what caused it but I've had a couple of the same errors show up before but didn't get stuck waiting so long for the tests to finish. I seem to be getting stuck on some tests too during the runs but I'm not sure if they're the same ones every time.

**Comment by juliagsy** [comment]
Hello! I'll check the changes and review soonest possible! Thanks for the good work!

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 1]
```diff
@@ -1473,6 +1473,50 @@ def test_cholesky(
...
+    num_positional_args,
+    native_array,
+    fn_name,
+    container_flags,
+    instance_method,
+    fw,
+):
+    dtype, x1, x2, _ = dtype_x1_x2
+    helpers.test_function(
+        input_dtypes=dtype,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1506]: please add a line: `ground_truth_backend=ground_truth_backend` above this argument

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 2]
```diff
@@ -1473,6 +1473,50 @@ def test_cholesky(
...
+        as_variable_flags=as_variable,
+        with_out=with_out,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        container_flags=container_flags,
+        instance_method=instance_method,
+        fw=fw,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1513]: please update `fw=fw` to `fw=backend_fw`

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 3]
```diff
@@ -1473,6 +1473,50 @@ def test_cholesky(
...
+        fn_name=fn_name,
+        x1=x1,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1515]: please add:
`on_device=on_device`
`test_gradients=True`

between the `fn_name` and `x1` arguments

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey there! The PR looks great so far! Just some small fixes on the test (it's probably due to our internal testing behaviour changes while you're in progress, therefore they're some newly added args which have yet to be added in your test function)

As usual, please request my review when you're done! Thanks for all the hard work!

**Comment by Hokchris** [comment]
Hey, just updated the test files. It got rid of the ERROR test outputs but I'm getting new failed runs instead. Most of them seem fine as they're situations which we want to fail (wrong input dimensions), but a couple have a datatype incongruence where numpy returns a float64 by default but torch returns a float32, and then the tensorflow output doesn't match with either one. Should I put a check based on what backend we're using and change the datatype that way or make a change to all the backends to have a default of float32 or float64 even when the datatype isn't supplied? Thanks!

**Comment by juliagsy** [comment]
Hey! I'll check the changes and get back to you as soon as possible on this! Thanks for the hard work!

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey! Apologies for the delay!

For the failing tests, we should only pass in input data with the accepted dimensions. If the default data generation helper functions don't allow this, you can create your own test helper at the top "helper" section of the test file which your test function is in!
An example: https://github.com/unifyai/ivy/blob/90107cdd390f287613bbd6dbe11e1a7c9c8db6cd/ivy_tests/test_ivy/test_frontends/test_torch/test_blas_and_lapack_ops.py#L11

As for the data type, the function should have a `@infer_dtype` decorator which handles dtype internally.
An example: https://github.com/unifyai/ivy/blob/90107cdd390f287613bbd6dbe11e1a7c9c8db6cd/ivy/functional/ivy/creation.py#L284

Hope this helps in fixing the failing tests! As usual, please request my review again when you're done! Thanks a lot for all the hard work!

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by Hokchris** [comment]
Hey! I think I'll have both changes done by end of day tomorrow, just need to fix a few more errors. Got the datatype infers fixed, just need to alter a new data generation helper function.

**Comment by juliagsy** [comment]
Hey! No worries! I've removed the stale label from the PR, take your time and just ping me when you think your PR is good for another review! Thanks!

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by Hokchris** [comment]
Hey! Just did some new changes to the branch again, had to make some changes to the tests to fix some of the backends and their datatypes. I'll try to get the aweights and fweights fixed soon, but most of the tests should be running fine. I've run into an issue of rounding with tensorflow as a backend however, getting this AssertionError:
[   25.600004    25.500004 -1433.3      -8243.5       9625.7     ] != [   25.600292    25.500195 -1433.2997   -8243.5       9625.699   ]

The numbers seem almost identical so I think the calculations are working but I'm guessing it might be either a datatype accuracy or atol/rtol (absolute/relative error tolerance) issue. Is there something I should do to fix this? Many thanks!

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 4]
```diff
@@ -1257,6 +1321,60 @@ def test_cholesky(
...
+        input_dtypes=dtype,
+        as_variable_flags=as_variable,
+        with_out=with_out,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        container_flags=container_flags,
+        instance_method=instance_method,
+        fw=backend_fw,
+        fn_name=fn_name,
+        on_device=on_device,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1366]: You may configure the tolerance by adding, eg, 

```
rtol=1e-01,
atol=1e-01,
```

as arguments in the `test_function`

**Review by juliagsy — COMMENTED** [review]
Hey @Hokchris ! Check the comment for suggestions, thanks!

**Comment by Hokchris** [comment]
Hey, just did some commits with a bunch of changes, should be pretty close to done now, I appreciate all the support! I'll try to get the finishing touches done tomorrow morning and I'll ping you once I think it's all done. I'm not sure if there's a new stale label on the branch again as I had another gap between commits but it should be all done tomorrow so I assume it's fine.

**Comment by Hokchris** [comment]
Hey, took a bit longer than I expected and I'm running into an error that I'm honestly not sure about at all. I've tested that the backends produce the same outputs with the given inputs but Jax and Torch have an AssertionError comparing the correct output to a seemingly random output (made by Tensorflow apparently), e.g. [0. 0.] != [-0.25 -0.25]. I've tested the same inputs with numpy separately and the outputs are correct but I'm getting a backend output from somewhere that I'm not seeing.

Also just to check, does the Ivy helpers "array_values" require another parameter other than min_value or abs_smallest_val to have a value range? I'm giving both 0.1 and 1 as minimum values but I'm getting an array with all 0's for some reason, which gives me a DivionByZeroError.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by juliagsy** [comment]
Hey! I'll review this soonest possible! Thanks a lot for your hard work!

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 5]
```diff
@@ -1257,6 +1346,50 @@ def test_cholesky(
...
+# cov
+@handle_test(
+    fn_tree="functional.ivy.cov",
+    dtype_x1_x2_cov=_get_dtype_value1_value2_cov(
+        available_dtypes=helpers.get_dtypes("numeric"),
+        min_num_dims=1,
+        max_num_dims=2,
+        min_dim_size=2,
+        max_dim_size=5,
+        min_value=-1e10,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1358]: It may be the configuration here:

The test error is:
`RuntimeError: cov(): weights sum to zero, can't be normalized`
(which may be the cause of the ZeroDivisionError)

You can try generating all positive values for example, setting the `min_value` to `0 or 1` (to prevent the probability of generating all 0s too) but keep the `max_value` as it is.

**Review by juliagsy — COMMENTED** [review]
Hey there! Apologies for the delay! Please check the suggestions in the comment below! Feel free to let me know if there are further issues! Happy to help! Thanks again for your hard work!

**Comment by Hokchris** [comment]
Hey again, I made the changes you recommended and tweaked some other values but still get some of the same issues, although not consistently (some runs one of the backends works, other times it gives the same ZeroDivisionError again).

I think the issue might be coming from the "create_args_kwargs" function inside the function_testing.py file (ivy_tests/test_ivy/helpers/function_testing.py) at around line number 1640. I think the datatype given to test_function is being applied to all inputs, even the aweights, and when the given datatype is an integer the aweights gets cast to the wrong datatype and rounded to 0. I tried setting the minimum value of aweights to be 1, but I still get some values below 1 for some reason. Should I do a separate datatype cast within the test_function parameters, or is there something else I should try instead?

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

File: ivy_tests/test_ivy/test_functional/test_core/test_linalg.py [hunk 6]
```diff
@@ -1258,6 +1347,50 @@ def test_cholesky(
...
+    test_flags,
+    backend_fw,
+    fn_name,
+    on_device,
+    ground_truth_backend,
+):
+    dtype, x1, x2, rowVar, bias, ddof, fweights, aweights = dtype_x1_x2_cov
+    helpers.test_function(
+        ground_truth_backend=ground_truth_backend,
+        input_dtypes=dtype,
```
Comment (juliagsy) [inline ivy_tests/test_ivy/test_functional/test_core/test_linalg.py:1378]: please specify the dtype for all arrays according to sequence, eg:

in this test,
1st arg is x1
2nd arg is x2
3rd arg is fweights
4th arg is aweights
(scalar like int or bool are skipped)

so, the input_dtypes should be `[dtype[0], dtype[0], 'int64', 'float64']`

this will prevent the test function from assuming all array arguments follow the same dtype and cast them all to that dtype

this solution should work! I've tried it out myself on `numpy` backend and the test is passing!

File: ivy/functional/backends/numpy/linear_algebra.py [hunk 7]
```diff
@@ -31,6 +31,40 @@ def cholesky(
...
+    bias: bool = False,
+    ddof: int = None,
+    fweights: np.ndarray = None,
+    aweights: np.ndarray = None,
+    dtype: Optional[np.dtype] = None,
+    out: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    x1, x2 = ivy.promote_types_of_inputs(x1, x2)
+
+    if fweights is not None:
```
Comment (juliagsy) [inline ivy/functional/backends/numpy/linear_algebra.py:50]: once the tests are configured correctly, implicit casting here shouldn't be necessary! unless it's related to the numpy implicit casting rule, you may have to tweak the tests a little to make sure array or dtype combinations follow the native conventions!

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey! Apologies for the delay! Took me some time to dive into the bug and investigate further, I've tried things out myself and suggested some possible solutions below, let me know if there are any further issues! Thanks for all the hard work!

**Comment by Hokchris** [comment]
Hey, was there a rebase done in the repo as this PR wants to merge 10,000 commits? The latest commit _ed9a037_ should fix the cov calculations, should I open a new PR?

**Comment by juliagsy** [comment]
Hey! Yes, there is a rebase, and yep, please create a new PR! Apologies for any inconvenience and thanks a lot!

**Comment by Hokchris** [comment]
Hey! Got it, made a new one already but I'll make another one tomorrow when its all setup. Thank you for all the support on this task, I very much appreciate it!

**Comment by juliagsy** [comment]
Sounds good! No worries at all and all the best to you too!
